### PR TITLE
ratelimit support query parameter match

### PR DESCRIPTION
**Pull request by StarryVae**
Signed-off-by: wangkai19 <wangkai19@corp.netease.com>

Commit Message: ratelimit support query parameter match
Additional Description: N/A
Risk Level: low
Testing: ut
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by StarryVae** [comment]
issue #23302 ratelimit supports query parameter match, here is the api refer to `HeaderValueMatch`, please review the api first, thanks. cc @wbpcode @mattklein123

**Comment by StarryVae** [comment]
ping @markdroth

**Comment by StarryVae** [comment]
ping @markdroth

**Comment by wbpcode** [comment]
Hi, @markdroth could you take a look for this new proto API when you have free time? Thanks.

**Comment by KBaichoo** [comment]
friendly ping @markdroth , it seems like this one is waiting on your input

**Comment by phlax** [comment]
@markdroth @envoyproxy/api-shepherds i think this one is waiting on api-shep approval

/wait-any

**Comment by StarryVae** [comment]
friendly ping @markdroth , cc @phlax @mattklein123 , thanks.

**Comment by wbpcode** [comment]
Seems that @markdroth is too busy and re-assign this PR.

**Comment by jmarantz** [comment]
@adisuissa @wbpcode can you take a look?

File: api/envoy/config/route/v3/route_components.proto [hunk 1]
```diff
@@ -2070,6 +2097,9 @@ message RateLimit {
 
       // Rate limit on masked remote address.
       MaskedRemoteAddress masked_remote_address = 10;
+
+      // Rate limit on the existence of query parameters.
+      QueryParameterValueMatch query_parameter_value_match = 11;
```
Comment (adisuissa) [inline api/envoy/config/route/v3/route_components.proto:2102]: A high-level question before going into this. Why isn't this a `rate_limit_descriptor` extension?
Making it an extension will avoid this from being linked if not needed.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this.
Left a high-level question, but other than that the `QueryParameterValueMatch` API LGTM.

File: api/envoy/config/route/v3/route_components.proto (hunk 1 above)
Comment (StarryVae) [reply api/envoy/config/route/v3/route_components.proto:2102]: @adisuissa thanks! but i am not familiar with `rate_limit_descriptor` extension, can it also support various types of matches(prefix, suffix, regex...) like QueryParameterValueMatch?

Comment (adisuissa) [reply api/envoy/config/route/v3/route_components.proto:2102]: There's an extension [here](https://github.com/envoyproxy/envoy/blob/a1606b267c27a6ffc057ac3abe7ae38734088265/source/extensions/rate_limit_descriptors/expr/config.h) that introduces a CEL rate-limiter.

That said, if you think there's a good reason for this to be integral part of Envoy's core, LMK.

Comment (StarryVae) [reply api/envoy/config/route/v3/route_components.proto:2102]: i think `rate_limit_descriptor` extension does not have all the capabilities of `QueryParameterMatcher` which supports different types of match for each query parameter.

**Comment by alyssawilk** [comment]
@adisuissa ping!

**Review by adisuissa — COMMENTED** [review]
Apologies for the late reply.
/lgtm api

**Comment by adisuissa** [comment]
@tyxia can you please take a first-pass at the code changes?

**Comment by tyxia** [comment]
/assign @tyxia 

Sure. I will take a look

**Comment by StarryVae** [comment]
@tyxia @wbpcode friendly ping.

**Comment by wbpcode** [comment]
sorry for the delay. I will take a look tomorrow.

File: test/common/router/router_ratelimit_test.cc [hunk 2]
```diff
@@ -1026,6 +1026,115 @@ TEST_F(RateLimitPolicyEntryTest, HeaderValueMatchHeadersPresent) {
...
+  setupTest(yaml);
+  Http::TestRequestHeaderMapImpl header{{":path", "/?x-parameter-name=not_same_value"}};
+
+  rate_limit_entry_->populateDescriptors(descriptors_, "", header, stream_info_);
+  rate_limit_entry_->populateLocalDescriptors(local_descriptors_, "", header, stream_info_);
+  EXPECT_TRUE(descriptors_.empty());
+  EXPECT_TRUE(local_descriptors_.empty());
+}
+
+TEST_F(RateLimitPolicyEntryTest, QueryParameterValueMatchNotPresent) {
```
Comment (tyxia) [inline test/common/router/router_ratelimit_test.cc:1094]: IMO, `QueryParameterValueMatchNotPresent` and `QueryParameterValueMatchPresent` are confusing, especially word `present`.

Basically,  `QueryParameterValueMatchNotPresent` is for case that `expect_not_match` and query parameters are indeed not matched (i.e., not_same_value != test_value)
`QueryParameterValueMatchPresent` is for case that `expect_not_match` but query parameters are matched.

I would suggest change name to `QueryParameterValueMatchExpectNoMatch` and `QueryParameterValueMatchExpectNoMatchFailed`. 
WDYT?

**Review by tyxia — CHANGES_REQUESTED** [review]
Thanks. Nice work! Mostly LGTM, just some nits.

Forward thinking, on the high level, [CEL expression](https://github.com/kyessenov/envoy/blob/a5389867fe536ab9676e61fe243c917f394934b7/source/extensions/filters/common/expr/context.cc#L109) could potentially be a good candidate to unify the matching methods here like `headerValue` and `QueryParameterValue`. But I think that is totally out of scope of this change and this PR matches existing styles/behaviors in rateLimit.

File: source/common/router/router_ratelimit.h [hunk 3]
```diff
@@ -179,6 +180,27 @@ class HeaderValueMatchAction : public RateLimit::DescriptorProducer {
...
+  bool populateDescriptor(RateLimit::DescriptorEntry& descriptor_entry,
+                          const std::string& local_service_cluster,
+                          const Http::RequestHeaderMap& headers,
+                          const StreamInfo::StreamInfo& info) const override;
+
+private:
+  const std::string descriptor_value_;
+  const std::string descriptor_key_;
+  const bool expect_match_;
+  std::vector<ConfigUtility::QueryParameterMatcherPtr> action_query_parameters_;
```
[the remark this answers is not in the document]
Comment (StarryVae) [reply source/common/router/router_ratelimit.h]: thanks, done

File: test/common/router/router_ratelimit_test.cc (hunk 2 above)
Comment (StarryVae) [reply test/common/router/router_ratelimit_test.cc]: thanks, done

**Review by tyxia — DISMISSED** [review]
Thanks!

File: source/common/router/router_ratelimit.cc [hunk 4]
```diff
@@ -241,6 +241,37 @@ bool HeaderValueMatchAction::populateDescriptor(RateLimit::DescriptorEntry& desc
...
+    : descriptor_value_(action.descriptor_value()),
+      descriptor_key_(!action.descriptor_key().empty() ? action.descriptor_key() : "query_match"),
+      expect_match_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(action, expect_match, true)),
+      action_query_parameters_(buildQueryParameterMatcherVector(action)) {}
+
+bool QueryParameterValueMatchAction::populateDescriptor(
+    RateLimit::DescriptorEntry& descriptor_entry, const std::string&,
+    const Http::RequestHeaderMap& headers, const StreamInfo::StreamInfo&) const {
+  Http::Utility::QueryParams query_parameters =
+      Http::Utility::parseQueryString(headers.getPathValue());
```
Comment (wbpcode) [inline source/common/router/router_ratelimit.cc:255]: nit: should we use `parseAndDecodeQueryString` here to decode escaped parameter value?

**Review by wbpcode — COMMENTED** [review]
LGTM overall. Only one nit comment. 

/wait-any

File: source/common/router/router_ratelimit.cc (hunk 4 above)
Comment (StarryVae) [reply source/common/router/router_ratelimit.cc:255]: yes, i think `parseAndDecodeQueryString` is more appropriate, thanks.

**Comment by StarryVae** [comment]
@adisuissa @wbpcode PTAL again, thanks.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by wbpcode** [comment]
ping cc @adisuissa for API tag.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by StarryVae** [comment]
ci is passed cc @wbpcode

**Pull request merged**